File: boardinghouse_app/lib/models/boarding_house.dart
import 'package:boardinghouse_app/models/Evaluate.dart';
import 'package:boardinghouse_app/models/boarding_house_type.dart';

import 'user.dart';
import 'util.dart';

class BoardingHouse {
  int? id;
  String? name;
  String? image; //???
  String? type; //???
  String? roomNumber;
  String? acreage;
  String? capacity;
  String? price;
  String? depositPrice;
  String? electricPrice;
  String? waterPrice;
  String? openTime;
  String? closeTime;
  String? description;
  String? address;
  // String? favourite;
  String? status;
  String? urlIamge;
  List<dynamic>? urlImages;
  int? userid;
  int? boardingHouseTypeId;

  BoardingHouseType? boardingHouseType;
  User? user;
  List<Utils>? utils;
  List<Evaluate>? evaluates;

  BoardingHouse({
    this.id,
    this.name,
    this.image, //???
    this.type, //??
    this.roomNumber,
    this.acreage,
    this.capacity,
    this.price,
    this.depositPrice,
    this.electricPrice,
    this.waterPrice,
    this.openTime,
    this.closeTime,
    this.description,
    this.address,
    // this.favourite,
    this.urlIamge,
    this.urlImages,
    this.status,
    this.userid,
    this.boardingHouseTypeId,
    this.boardingHouseType,
    this.user,
    this.utils,
    this.evaluates,
  });

  factory BoardingHouse.fromJson(Map<String, dynamic> json) {
    return BoardingHouse(
      id: json['id'],
      name: json['name'],
      image: json['featured_image_id'], //???
      type: json['type'], //???
      roomNumber: json['room_number'],
      acreage: json['acreage'],
      capacity: json['capacity'],
      price: json['price'],
      depositPrice: json['deposit_price'],
      electricPrice: json['electric_price'],
      waterPrice: json['water_price'],
      // openTime: json['open_time'],
      // closeTime: json['close_time'],
      description: json['description'],
      address: json['address'],
      urlIamge: json['url_featured_image'],
      urlImages: json['url_images'],
      // favourite: json['favourite'],
      status: json['status'],
      userid: json['user_id'],
      boardingHouseTypeId: json['boarding_house_type_id'],
      boardingHouseType: BoardingHouseType(
          id: json['boarding_house_type']['id'],
          name: json['boarding_house_type']['name']),
      user: User(
        id: json['user']['id'],
        userName: json['user']['user_name'],
        urlAvatar: json['user']['url_avatar'],
      ),
      utils:
          List<Utils>.from(json['utils'].map((util) => Utils.fromJson(util))),
      evaluates: List<Evaluate>.from(
        json['evaluates'].map((evaluate) => Evaluate.fromJson(evaluate)),
      ),
    );
  }
}

File: boardinghouse_backend/app/Models/BoardingHouse.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Facades\Storage;

class BoardingHouse extends Model
{
    use HasFactory;

    protected $appends = ['url_featured_image', 'url_images', 'utils', 'boarding_house_type', 'user', 'evaluates', 'user_email', 'user_phone'];
    protected $fillable = [
        'type',
        'name',
        'images',
        'featured_image',
        'room_number',
        'acreage',
        'capacity',
        'price',
        'deposit_price',
        'electric_price',
        'water_price',
        'open_time',
        'close_time',
        'description',
        'address',
        'favourite',
        'status',
        'user_id',
        'published_at',
        'boarding_house_type_id'
    ];

    protected $casts = ['images' => 'array'];

    public function boarding_house_type(): BelongsTo
    {
        return $this->belongsTo(BoardingHouseType::class, 'boarding_house_type_id');
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class, 'user_id');
    }


    public function evaluates(): HasMany
    {
        return $this->hasMany(Evaluate::class, 'boarding_house_id');
    }

    public function posts(): HasMany
    {
        return $this->hasMany(Post::class, 'boarding_house_id');
    }

    public function utils(): BelongsToMany
    {
        return $this->belongsToMany(Util::class, 'util_boarding_houses', 'boarding_house_id', 'util_id')->withTimestamps();
    }
    public function favourites(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'favourite_boarding_houses', 'boarding_house_id', 'user_id')->withTimestamps();
    }

    public function getUrlFeaturedImageAttribute()
    {
        if (empty($this->featured_image) || !$this->featured_image) return;

        return url(Storage::url($this->featured_image));
    }

    public function getUtilsAttribute()
    {
        return $this->utils()->get();
    }

    public function getBoardingHouseTypeAttribute()
    {
        return BoardingHouseType::findOrFail($this->boarding_house_type_id);
    }

    public function getUserAttribute()
    {
        return User::findOrFail($this->user_id);
    }

    public function getEvaluatesAttribute()
    {
        return $this->evaluates()->get();
    }

    public function getUrlImagesAttribute()
    {
        if (empty($this->images)) return;
        $urlImages = [];

        foreach ($this->images as $image) {
            $urlImages[] = url(Storage::url($image));
        }

        return $urlImages;
    }

    public function getUserEmailAttribute()
    {
        return $this->user->email;
    }
    public function getUserPhoneAttribute()
    {
        return $this->user->phone;
    }
}
